File: WorldFeed/Server/src/WorldFeed.Common/RateLimit/IRateLimitUpdater.cs
namespace WorldFeed.Common.RateLimit
{
    using System.Collections.Generic;
    using System.Threading.Tasks;

    using WorldFeed.Common.Models.Authentication;
    using WorldFeed.Common.Public.Models.Authentication;

    public interface IRateLimitUpdaterFactory
    {
        IRateLimitUpdater Create(IRateLimitCacheManager rateLimitCacheManager);
    }

    /// <summary>
    /// Update the rate limit cached information.
    /// </summary>
    public interface IRateLimitUpdater
    {
        /// <summary>
        /// Inform the updater a specific query has been executed with a specific set of credentials.
        /// </summary>
        Task QueryExecutedAsync(string query, ITwitterCredentials credentials, int numberOfRequests = 1);

        /// <summary>
        /// Inform the updater a specific query has been executed with a specific set of credentials.
        /// </summary>
        Task QueryExecutedAsync(string query, ITwitterCredentials credentials, Dictionary<string, IEnumerable<string>> rateLimitHeaders);

        /// <summary>
        /// Inform that you want to query rate limits to be set to 0.
        /// </summary>
        Task ClearRateLimitsForQueryAsync(string query, IReadOnlyTwitterCredentials credentials);
    }
}

File: Sharebook/Client/src/app/core/Core/RateLimit/IRateLimitUpdater.ts
import {inject, Inject, InjectionToken} from "@angular/core";

import {IRateLimitCacheManager} from "./IRateLimitCacheManager";
import {IReadOnlyTwitterCredentials} from "../Models/Authentication/ReadOnlyTwitterCredentials";
import {ITwitterCredentials} from "../../Public/Models/Authentication/TwitterCredentials";
import {RateLimitUpdater, RateLimitUpdaterFactory} from "../../../Tweetinvi.Credentials/RateLimit/RateLimitUpdater";
import Dictionary from "typescript-dotnet-commonjs/System/Collections/Dictionaries/Dictionary";

export interface IRateLimitUpdaterFactory {
  create(rateLimitCacheManager: IRateLimitCacheManager): IRateLimitUpdater;
}

export const IRateLimitUpdaterFactoryToken = new InjectionToken<IRateLimitUpdaterFactory>('IRateLimitUpdaterFactory', {
  providedIn: 'root',
  factory: () => new RateLimitUpdaterFactory(),
});


// Update the rate limit cached information.
export interface IRateLimitUpdater {
  // Inform the updater a specific query has been executed with a specific set of credentials.
  queryExecutedAsync(query: string, credentials: ITwitterCredentials, numberOfRequests): Promise<void>;

  // Inform the updater a specific query has been executed with a specific set of credentials.
  queryExecutedAsync(query: string, credentials: ITwitterCredentials, rateLimitHeaders: Dictionary<string, Array<string>>): Promise<void>;

  // Inform that you want to query rate limits to be set to 0.
  clearRateLimitsForQueryAsync(query: string, credentials: IReadOnlyTwitterCredentials): Promise<void>;
}

export const IRateLimitUpdaterToken = new InjectionToken<IRateLimitUpdater>('IRateLimitUpdater', {
  providedIn: 'root',
  factory: () => new RateLimitUpdater(inject(RateLimitUpdater)),
});
